File: leetcode/0-300/0-50/20.Valid Parentheses.cc
// https://leetcode.com/problems/valid-parentheses/
 
class Solution {
private:
    stack<char> l_parentheses;
private:
    bool isLefty(char p)
    {
        return p == '[' || p == '(' || p == '{';  
    }
    
    bool matches(char lhs, char rhs)
    {
        return lhs == '[' && rhs == ']'
            || lhs == '(' && rhs == ')'
            || lhs == '{' && rhs == '}';
    }
public:
    bool isValid(string s) {
        for(char c: s) {
            if (isLefty(c)) {
                l_parentheses.push(c);
            } else if(!l_parentheses.empty() && matches(l_parentheses.top(), c)) {
                l_parentheses.pop();
            } else {
                return false;
            }
        }
        
        return l_parentheses.empty();
    }
};

File: leetcode/0-300/0-50/20.Valid Parentheses.rs
// https://leetcode.com/problems/valid-parentheses

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = vec!();
        for ch in s.chars() {
            match ch {
                '(' | '[' | '{' => stack.push(ch),
                ')' | ']' | '}' => {
                    if let Some(last) = stack.last() {
                        if (*last == '(' && ch != ')') || 
                            (*last == '[' && ch != ']') || 
                            (*last == '{' && ch != '}') 
                        {
                            return false;
                        }    
                        stack.pop();
                        continue;
                    }

                    return false
                }
                _ => continue,
            }
        }
        return stack.len() == 0
    }
}

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = vec!();
        for ch in s.chars() {
            match ch {
                '(' => stack.push(')'),
                '[' => stack.push(']'),
                '{' => stack.push('}'),
                ')' | ']' | '}' => if Some(ch) != stack.pop() { return false; }
                _ => continue,
            }
        }
        return stack.is_empty()
    }
}